File: backend/src/services/crmAssistant.ts
import Anthropic from "@anthropic-ai/sdk";
import { config } from "../config";
import type { LeadStatus } from "./leadScoring";
import type { StudentRow } from "./db";
import { explainApplicationStatus } from "./applicationStatus";
import { buildDocumentChecklist, type DocumentRecord } from "./documentChecklist";

const client = new Anthropic({ apiKey: config.anthropicApiKey });

// Named, documented thresholds — same rubric philosophy as leadScoring.ts:
// a Hot lead going quiet is far more time-sensitive than a Cold one.
export const INACTIVITY_THRESHOLD_DAYS: Record<LeadStatus, number> = {
  Hot: 2,
  Warm: 5,
  Cold: 14,
};

const MS_PER_DAY = 1000 * 60 * 60 * 24;

export interface InactivityCheckResult {
  isInactive: boolean;
  daysSinceContact: number | null; // null = never contacted at all
  thresholdDays: number;
}

/**
 * Pure, deterministic — no I/O, no LLM call. A student who has reached the
 * terminal "enrolled" status is never flagged: there's no further pipeline
 * movement to chase.
 */
export function checkInactivity(student: StudentRow, now: Date = new Date()): InactivityCheckResult {
  const thresholdDays = INACTIVITY_THRESHOLD_DAYS[student.status];
  if (student.application_status === "enrolled") {
    return { isInactive: false, daysSinceContact: null, thresholdDays };
  }

  const referenceIso = student.last_contacted_at ?? student.created_at;
  const daysSince = Math.floor((now.getTime() - new Date(referenceIso).getTime()) / MS_PER_DAY);

  return {
    isInactive: daysSince >= thresholdDays,
    daysSinceContact: student.last_contacted_at ? daysSince : null,
    thresholdDays,
  };
}

export interface FollowupSuggestion {
  timing: string;
  suggestedAction: string;
  messageTemplate: string;
}

const FOLLOWUP_SCHEMA = {
  type: "object",
  properties: {
    suggestedAction: {
      type: "string",
      description: "One or two sentences telling the counsellor exactly what to do right now (call, email, chase a document, etc).",
    },
    messageTemplate: {
      type: "string",
      description: "A short, ready-to-send message addressed directly to the student, personalized to their specific situation. Plain text, no subject line.",
    },
  },
  required: ["suggestedAction", "messageTemplate"],
  additionalProperties: false,
} as const;

const FOLLOWUP_PERSONA = `You are an assistant to a study-abroad counsellor at AIEC Global. You are given one student's CRM record and must suggest what the counsellor should do next to re-engage them.

Do not invent facts not present in the STUDENT RECORD below (no fee/policy/visa claims). Be specific to this student — reference their actual situation (status, missing documents, how long it's been) rather than writing something generic that could apply to anyone.`;

/**
 * On-demand only (never auto-run per row on every dashboard load) — an
 * explicit "regenerate suggestion" action, same cost-control reasoning as
 * the rest of this codebase (rule-based where deterministic suffices,
 * Claude only where personalization genuinely adds value).
 */
export async function generateFollowupSuggestion(
  student: StudentRow,
  inactivity: InactivityCheckResult,
): Promise<FollowupSuggestion> {
  const statusInfo = explainApplicationStatus(student.application_status);
  const documents = student.documents ? (JSON.parse(student.documents) as DocumentRecord[]) : [];
  const checklist = buildDocumentChecklist(documents);

  const timing = inactivity.isInactive
    ? `Overdue now — contact within 24 hours (${student.status} leads are expected to be contacted every ${inactivity.thresholdDays} day(s))`
    : `On track — next contact due in ${inactivity.thresholdDays - (inactivity.daysSinceContact ?? 0)} day(s)`;

  const recordBlock = [
    `STUDENT RECORD:`,
    `- Name: ${student.name}`,
    `- Lead classification: ${student.status}`,
    `- Application status: ${statusInfo.label} — ${statusInfo.description}`,
    `- Days since last contact: ${inactivity.daysSinceContact ?? "never contacted"}`,
    `- Preferred country: ${student.preferred_country ?? "not specified"}`,
    `- Career goals: ${student.career_goals || "not specified"}`,
    `- Missing documents: ${checklist.missing.length > 0 ? checklist.missing.join(", ") : "none"}`,
    `- Documents needing attention: ${checklist.needsAttention.length > 0 ? checklist.needsAttention.join(", ") : "none"}`,
    `- Counsellor notes: ${student.counsellor_notes || "(none)"}`,
  ].join("\n");

  const response = await client.messages.create({
    model: config.claudeModel,
    max_tokens: 512,
    thinking: { type: "adaptive" },
    output_config: {
      effort: "low",
      format: { type: "json_schema", schema: FOLLOWUP_SCHEMA },
    },
    system: [
      { type: "text", text: FOLLOWUP_PERSONA, cache_control: { type: "ephemeral" } },
      { type: "text", text: recordBlock },
    ],
    messages: [{ role: "user", content: "Suggest my next action and a message I can send this student." }],
  });

  const textBlock = response.content.find((block) => block.type === "text");
  const raw = textBlock && textBlock.type === "text" ? textBlock.text : "{}";
  const parsed = JSON.parse(raw) as { suggestedAction: string; messageTemplate: string };

  return { timing, suggestedAction: parsed.suggestedAction, messageTemplate: parsed.messageTemplate };
}

File: backend-py/app/services/crm_assistant.py
import json
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from app.services.application_status import explain_application_status
from app.services.document_checklist import DOCUMENT_LABELS, build_document_checklist


def _join_readable(items: list[str]) -> str:
    """["a"] -> "a"; ["a","b"] -> "a and b"; ["a","b","c"] -> "a, b, and c" —
    a document list joined with bare commas ("citizenship, marksheet,
    ielts_certificate") reads like a raw data dump, not a sentence a human
    wrote."""
    if len(items) <= 1:
        return items[0] if items else ""
    if len(items) == 2:
        return f"{items[0]} and {items[1]}"
    return f"{', '.join(items[:-1])}, and {items[-1]}"


def _document_labels(doc_types: list[str]) -> list[str]:
    return [DOCUMENT_LABELS[t] for t in doc_types]

# Named, documented thresholds — same rubric philosophy as lead_scoring.py: a
# Hot lead going quiet is far more time-sensitive than a Cold one.
INACTIVITY_THRESHOLD_DAYS = {"Hot": 2, "Warm": 5, "Cold": 14}


@dataclass
class InactivityCheckResult:
    isInactive: bool
    daysSinceContact: int | None  # None = never contacted at all
    thresholdDays: int


def _parse_iso(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def check_inactivity(student: Any, now: datetime | None = None) -> InactivityCheckResult:
    """Pure, deterministic — no I/O, no LLM call. A student who has reached the
    terminal "enrolled" status is never flagged."""
    now = now or datetime.now(timezone.utc)
    threshold_days = INACTIVITY_THRESHOLD_DAYS[student["status"]]
    if student["application_status"] == "enrolled":
        return InactivityCheckResult(isInactive=False, daysSinceContact=None, thresholdDays=threshold_days)

    reference_iso = student["last_contacted_at"] or student["created_at"]
    days_since = (now - _parse_iso(reference_iso)).days

    return InactivityCheckResult(
        isInactive=days_since >= threshold_days,
        daysSinceContact=days_since if student["last_contacted_at"] else None,
        thresholdDays=threshold_days,
    )


@dataclass
class FollowupSuggestion:
    timing: str
    suggestedAction: str
    messageTemplate: str


# No LLM here — suggestion + message are built from the same STUDENT RECORD
# fields the Claude version was given, via fixed templates instead of
# free-form writing. Less natural-sounding, but every word is traceable to a
# specific field, same anti-invention principle as the rest of this codebase.


def generate_followup_suggestion(student: Any, inactivity: InactivityCheckResult) -> FollowupSuggestion:
    """On-demand only (never auto-run per row on every dashboard load) —
    matches the original's "explicit regenerate action" behavior, though the
    cost reasoning that motivated that no longer applies without an LLM call."""
    status_info = explain_application_status(student["application_status"])
    documents = json.loads(student["documents"]) if student["documents"] else []
    checklist = build_document_checklist(documents)

    if inactivity.isInactive:
        timing = (
            f"Overdue now — contact within 24 hours ({student['status']} leads are expected to be "
            f"contacted every {inactivity.thresholdDays} day(s))"
        )
    else:
        remaining = inactivity.thresholdDays - (inactivity.daysSinceContact or 0)
        timing = f"On track — next contact due in {remaining} day(s)"

    name = student["name"]

    if checklist.missing:
        missing_labels = _join_readable(_document_labels(checklist.missing))
        action = f"Ask {name} to upload the missing document(s): {missing_labels}."
    elif checklist.needsAttention:
        attention_labels = _join_readable(_document_labels(checklist.needsAttention))
        action = f"Ask {name} to re-upload the flagged document(s): {attention_labels}."
    elif inactivity.isInactive:
        action = f"Reach out to {name} — it's been {inactivity.daysSinceContact if inactivity.daysSinceContact is not None else 'a while'} day(s) since last contact."
    else:
        action = f"No urgent action needed for {name} right now — next contact is on track."

    message_lines = [f"Hi {name},"]
    if checklist.missing:
        missing_labels = _document_labels(checklist.missing)
        pronoun = "it" if len(missing_labels) == 1 else "them"
        message_lines.append(
            f"We're still waiting on your {_join_readable(missing_labels)} — could you upload {pronoun} when you get a chance?"
        )
    elif checklist.needsAttention:
        attention_labels = _document_labels(checklist.needsAttention)
        message_lines.append(
            f"We noticed an issue with your {_join_readable(attention_labels)} — could you re-upload a clearer copy?"
        )
    else:
        message_lines.append(
            f"Just checking in on your application ({status_info.label.lower()}) — let us know if you have any questions."
        )
    message_lines.append("Best, AIEC Global")

    return FollowupSuggestion(timing=timing, suggestedAction=action, messageTemplate="\n".join(message_lines))
